### ISSUE TITLE
[BUG] `warn_dependency_requirement_mismatches` emits spurious warnings for requirements mismatches over model logged with uv detection

### ISSUE THREAD

**Issue by WeichenXu123** [mlflow/mlflow#21765]
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Issues Policy acknowledgement

- [x] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### MLflow version

- Client: MLflow master

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOS
- **Python version**: 3.10

### Describe the problem

`warn_dependency_requirement_mismatches` emits spurious warnings for requirements mismatches over model logged with uv detection:

For model logged with uv detection, the requirements.txt file (generated by `uv export`) might contain environment marker like:

```
numpy==2.2.6; python_full_version < "3.11"
numpy==2.4.3; python_full_version >= "3.11"
nvidia-nccl-cu12==2.29.7; sys_platform == "linux"
```

When running model inference with uv env-manager, `warn_dependency_requirement_mismatches` does not handle this case well and raise spurious warnings like:

```
 - numpy (current: 2.2.6, required: numpy==2.4.3; python_full_version >= "3.11")
 - nvidia-nccl-cu12 (current: uninstalled, required: nvidia-nccl-cu12==2.29.7; sys_platform == "linux")
```

### Tracking information

```shell
REPLACE_ME
```

### Code to reproduce issue

Environment: my MacOS with **python 3.10**

1. checkout mlflow master,
2. then run `pip install -e .` ,
3. then enter `examples/uv-dependency-management`,
4. then run `python log_model.py`

5. create a python script at `/tmp/p1.py` as follows:

```
import mlflow
from mlflow.models import predict
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

iris = load_iris()
X_train, X_test, y_train, y_test = train_test_split(
    iris.data, iris.target, test_size=0.2, random_state=42
)
model_uri = <model uri to the model logged by example_groups_and_extras>
predict(
    model_uri=model_uri,
    input_data=X_test,
    env_manager="uv",
)
```
run `python /tmp/p1`

and it emits warning like
```
2026/03/17 22:10:44 WARNING mlflow.utils.requirements_utils: Detected one or more mismatches between the model's dependencies and the current Python environment:
 - colorama (current: uninstalled, required: colorama==0.4.6; sys_platform == "win32")
 - contourpy (current: 1.3.2, required: contourpy==1.3.3; python_full_version >= "3.11")
 - greenlet (current: uninstalled, required: greenlet==3.3.2; platform_machine == "AMD64" or platform_machine == "WIN32" or platform_machine == "aarch64" or platform_machine == "amd64" or platform_machine == "ppc64le" or platform_machine == "win32" or platform_machine == "x86_64")
 - numpy (current: 2.2.6, required: numpy==2.4.3; python_full_version >= "3.11")
 - nvidia-nccl-cu12 (current: uninstalled, required: nvidia-nccl-cu12==2.29.7; sys_platform == "linux")
 - pywin32 (current: uninstalled, required: pywin32==311; sys_platform == "win32")
 - scikit-learn (current: 1.7.2, required: scikit-learn==1.8.0; python_full_version >= "3.11")
 - scipy (current: 1.15.3, required: scipy==1.17.1; python_full_version >= "3.11")
 - waitress (current: uninstalled, required: waitress==3.0.2; sys_platform == "win32")
 - xgboost (current: uninstalled, required: xgboost==3.2.0)
```

### Stack trace

```
REPLACE_ME
```

### Other info / logs

```
REPLACE_ME
```

### What component(s) does this bug affect?

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow model serving, deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting
- [ ] `area/docs`: MLflow documentation pages

**Comment by FocusMode-coder** [comment]
Hey! I may be able to help with this bug.

I build custom fixes for Python projects and automation systems.

If you'd like help resolving this issue quickly feel free to reach out.

Happy to take a look 👍

**Comment by WeichenXu123** [comment]
oh I just assigned this to @debu-sinha . If he doesn't have time, I'll assign it to you :)

**Comment by debu-sinha** [comment]
PR #21769 addresses this. The fix adds a pre-filter in `warn_dependency_requirement_mismatches` that evaluates environment markers before version checking.

Tested end-to-end on Python 3.10/macOS against PR #21750's branch with uv-style requirements containing markers for both Python 3.10 and 3.11:

```
Input requirements (simulating uv export):
  numpy==2.2.6; python_full_version < "3.11"
  numpy==2.4.3; python_full_version >= "3.11"
  nvidia-nccl-cu12==2.29.7; sys_platform == "linux"
  colorama==0.4.6; sys_platform == "win32"
  scikit-learn==1.7.2; python_full_version < "3.11"
  scikit-learn==1.8.0; python_full_version >= "3.11"

Result: zero spurious warnings. Non-matching markers (python >= 3.11,
sys_platform == "linux", sys_platform == "win32") correctly filtered.
Matching markers (python < 3.11) evaluated normally.
```

**Comment by harupy** [comment]
@debu-sinha Were you able to reproduce the warning?

**Comment by debu-sinha** [comment]
@harupy No, I was not able to reproduce the spurious warning on Python 3.10/macOS (arm64). The existing marker check in `_check_requirement_satisfied` (line 624) correctly skips requirements with non-matching markers in my environment.

I tested both directly calling `warn_dependency_requirement_mismatches` with the exact requirement strings from the issue, and also tested against PR #21750's branch. In both cases, markers like `python_full_version >= "3.11"` and `sys_platform == "linux"` evaluated correctly and were skipped.

@WeichenXu123 Could you share the exact Python version and platform where you saw the warning? I want to make sure I'm testing under the same conditions. If the issue is environment-specific, the defensive pre-filter in #21769 would still help as a safety net, but I'd like to understand the root cause.

**Comment by WeichenXu123** [comment]
Python 3.10.19

the model's uvlock file:

https://gist.github.com/WeichenXu123/0e9b653548d5ae15af7f4dd9cb36e10b 

you can use it to reproduce.

the fix https://github.com/mlflow/mlflow/pull/21768 does not address the issue in my test.

**Comment by debu-sinha** [comment]
@WeichenXu123 Thanks for the uv.lock file. I ran `uv export --no-hashes --frozen` against it and tested the exact output format (single-quoted markers like `python_full_version >= '3.11'`) with `warn_dependency_requirement_mismatches` on Python 3.10.19/macOS arm64.

All markers evaluate correctly. Zero spurious warnings. The existing check at line 624 handles both single and double quoted markers.

```
numpy==2.2.6 ; python_full_version < '3.11'    -> marker=True  (checked, matches)
numpy==2.4.3 ; python_full_version >= '3.11'   -> marker=False (skipped)
nvidia-nccl-cu12==2.29.7 ; sys_platform == 'linux' -> marker=False (skipped)
scikit-learn==1.8.0 ; python_full_version >= '3.11' -> marker=False (skipped)
```

I cannot reproduce the warning. Could you share the exact reproduction steps you used? Specifically:
1. Which command logged the model (the example script, or something else)?
2. Which command triggered the warning (`mlflow.models.predict` with `env_manager="uv"`)?
3. Was the warning from the parent process or from inside the uv subprocess?

There might be a different code path or a requirements.txt transformation step I'm not testing.

**Comment by WeichenXu123** [comment]
I got the reason:

the issue has been fixed in https://github.com/mlflow/mlflow/pull/21098
related added line
```
    if req.marker and not req.marker.evaluate():
        return None
```

but the fix only went into python >= 3.11 (currently we only have python 3.11.0rc0 available)

then the project https://github.com/mlflow/mlflow/blob/6157243f2ef261f0abe1bb217178bb2ce271ae8b/examples/uv-dependency-management/pyproject.toml#L6 generates uv.lock with mlflow version 3.10.1

then when log models with uv project detection, it generates mlflow==3.10.1 in the model's uv.lock file.
then in model prediction python environment , mlflow 3.10.1 is installed , the mlflow version does not have the above fix, so it emits the spurious warning

**Comment by WeichenXu123** [comment]
so that we just need to wait for mlflow 3.11.0 release. then it is addressed
